File: CPP/ByRound/1930/1930_F_Maximize_the_Difference.cpp
// Problem: CF 1930 F - Maximize the Difference
// https://codeforces.com/contest/1930/problem/F

/*
 * Problem: Maximize the Difference
 * Algorithm/Technique: Bit manipulation with trie-like structure
 *
 * Time Complexity: O(q * log n), where q is the number of queries and n is the size of the array.
 * Space Complexity: O(n), due to the boolean flag array used to track reachable values.
 *
 * This solution uses a clever bit manipulation trick to efficiently compute the maximum
 * difference between the max and min of the array after each query.
 * The key idea is to maintain two sets:
 *   - T0: contains all possible values that can be achieved by OR-ing elements with some x.
 *   - T1: contains all current elements in the array.
 * Then, we try to find the largest value that is present in both sets to compute the maximum difference.
 */
#include <stdio.h>
#include <iosfwd>

using namespace std;
struct VG{
  bool fl[1 << 22]; // Boolean array to mark reachable values (bit manipulation optimization)
  void co(int x) {
    if (fl[x])
      return;
    fl[x] = 1;
    for (int p = x; p; p -= (p & -p)) // Clear lowest set bit iteratively
      co(x ^ (p & -p)); // Toggle bits of x to find all combinations
  }
  void init(int n) {
    for (int i = 0; i < n; ++i)
      fl[i] = 0;
  }
} T1, T0;
void sol() {
  int n, q;
  scanf("%d%d", &n, &q);
  int e = 1, z = 0;
  while (e < n)
    e <<= 1, z++; // Find smallest power of 2 >= n and calculate bit length z
  T0.init(e);
  T1.init(e);
  int ans = 0;
  for (int i = 1, a; i <= q; ++i) {
    scanf("%d", &a);
    (a += ans) %= n; // Decrypt the value using previous output
    T0.co(a ^ ((1 << z) - 1)); // Mark all possible XOR values with all bits set
    T1.co(a); // Mark current value
    int st = 0;
    for (int j = z - 1; j >= 0; --j)
      if (T0.fl[st | (1 << j)] && T1.fl[st | (1 << j)]) // Try to build result greedily
        st |= 1 << j;
    printf("%d\n", ans = st); // Output result for this query
  }
}
int main() {
  int T;
  scanf("%d", &T);
  while (T--)
    sol();
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/F/1930_F_Maximize_the_Difference.py
# Problem: CF 1930 F - Maximize the Difference
# https://codeforces.com/contest/1930/problem/F

"""
Algorithm: Maximize the Difference
Purpose: For each query, append a value to an array and compute the maximum difference 
         between max and min elements in the array over all possible bitwise OR operations 
         with a non-negative integer x.
         
Time Complexity: O(q * u * log u) where u = 2^(bit_length(n))
Space Complexity: O(u) where u = 2^(bit_length(n))

Techniques:
- Bit manipulation and bitmasking
- BFS-like traversal on bits
- Preprocessing and tracking of seen states
"""

import sys

input = sys.stdin.readline

out = []

t = int(input())
for _ in range(t):
    n, q = map(int, input().split())
    l = list(map(int, input().split()))

    # Determine the bit length needed to represent n
    ul = n.bit_length()
    u = 1 << ul  # 2^ul
    ms = u - 1   # Mask to get all bits set

    # f0 and f1 track the states visited in two different iterations
    f0 = [0] * u
    f1 = [0] * u

    v = 0
    res = []
    ans = 0

    for i in range(q):
        # Decrypt v from encrypted input
        v += l[i]
        v %= n

        # First phase: Traverse all submasks of v and update ans based on f1
        st = [v]
        while st:
            u = st.pop()
            if f0[u]:
                continue

            if f1[u]:
                ans = max(ans, u)
            f0[u] = 1

            # For each bit, push the submask (v with that bit cleared)
            for b in range(ul):
                st.append(u & (~(1 << b)))

        # Set v to complement of v (masked to keep only up to ul bits)
        v ^= ms
        st = [v]
        while st:
            u = st.pop()
            if f1[u]:
                continue

            if f0[u]:
                ans = max(ans, u)
            f1[u] = 1

            # For each bit, push the submask (v with that bit cleared)
            for b in range(ul):
                st.append(u & (~(1 << b)))

        v = ans
        res.append(v)
    out.append(" ".join(map(str, res)))
print("\n".join(out))


# https://github.com/VaHiX/CodeForces/